classes/migration/upgrade/v3_4_0/I7167_RemoveDuplicatedUserSettingsAndDeprecatedFields.inc.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+/**
+ * @file classes/migration/upgrade/v3_4_0/I7167_RemoveDuplicatedUserSettingsAndDeprecatedFields.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class I7167_RemoveDuplicatedUserSettingsAndDeprecatedFields
+ * @brief Describe upgrade/downgrade Removes duplicated user_settings, as well as the deprecated fields "assoc_id" and "assoc_type".
+ */
+
+namespace PKP\migration\upgrade\v3_4_0;
+
+use Exception;
+use Illuminate\Database\MySqlConnection;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+use PKP\install\DowngradeNotSupportedException;
+use PKP\migration\Migration;
+
+class I7167_RemoveDuplicatedUserSettingsAndDeprecatedFields extends Migration
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        if (!Schema::hasColumn('user_settings', 'assoc_id') && !Schema::hasColumn('user_settings', 'assoc_type')) {
+            return;
+        }
+
+        $isMysql = DB::connection() instanceof MySqlConnection;
+        $deleteFrom = 'DELETE FROM user_settings s USING';
+        $joinFilter = 'WHERE';
+        if ($isMysql) {
+            // The optimizer might cut-off the sub-queries/closures below and make MySQL fail to delete from the table which is selecting from
+            DB::unprepared("SET optimizer_switch = 'derived_merge=off'");
```

Comment:
@jonasraoni, can you document this a little more in the comment? e.g. is there a link to information about it, and conditions to replicate? We may need to consider this problem elsewhere (e.g. https://github.com/pkp/pkp-lib/pull/7178/files#r755415915) and I'd like to make sure we're treating joins with updates/deletes systematically.

Reply:
Ah, this problem is old, it happens in MySQL when you try to `DELETE`/`UPDATE` and there's a sub-query using the same target table.
It was possible to fix it by enclosing the sub-query with another `(SELECT * FROM (original subquery) AS fix`, but it doesn't work in newer versions of MySQL, which detect the useless sub-query and removes it :D

Then, disabling this optimizer setting, avoids this behavior...

In case you'd like to see something less hacky, but slower, I could create a temporary table to store the IDs which should be removed.

Reply:
Now that I'm re-watching the query, maybe it's possible to solve the problem without a sub-query, I'll check 🤔

Reply:
Yeah, it worked, I compared the outputs and got the same result, I'll update the PR tomorrow with this:
```sql
DELETE s
FROM user_settings s
INNER JOIN user_settings duplicated
	ON s.setting_name = duplicated.setting_name
    AND s.user_id = duplicated.user_id
    AND s.locale = duplicated.locale
    AND (
		COALESCE(s.assoc_type, -9999) <> COALESCE(duplicated.assoc_type, -9999)
        OR COALESCE(s.assoc_id, -9999) <> COALESCE(duplicated.assoc_id, -9999)
	)
LEFT JOIN user_settings best
	ON best.setting_name = duplicated.setting_name
	AND best.user_id = duplicated.user_id
	AND best.locale = duplicated.locale
    AND COALESCE(best.assoc_id, 9999) < COALESCE(s.assoc_id, 9999)
    AND COALESCE(best.assoc_type, 9999) < COALESCE(s.assoc_id, 9999)
WHERE best.user_id IS NOT NULL
```

Reply:
I worry that this may be one of those things that pass Travis and local testing but result in errors on other versions of MySQL/PostgreSQL that we're not testing -- is the original issue documented somewhere?

Reply:
Yeah, it's documented here:
https://dev.mysql.com/doc/refman/8.0/en/delete.html#idm46019841632784:~:text=You%20cannot%20delete%20from%20a%20table%20and%20select%20from%20the%20same%20table%20in%20a%20subquery.

Reply:
Ok! I've updated the query to avoid using a sub-query.